replicatelogs.go

```diff
@@ -99,58 +123,99 @@ changes are read from standard input.`,
 			}
 			progress := newProgressor(cCtx, "tenants", len(changes))
 
-			var wg sync.WaitGroup
-			errChan := make(chan error, len(changes)) // buffered so it doesn't block
-
-			for _, change := range changes {
-				wg.Add(1)
-				go func(change TenantMassif, errChan chan<- error) {
-					defer wg.Done()
-					defer progress.Completed()
-
-					replicator, err := NewVerifiedReplica(cCtx, cmd.Clone())
-					if err != nil {
-						errChan <- err
-						return
-					}
-					endMassif := uint32(change.Massif)
-					startMassif := uint32(0)
-					if cCtx.IsSet("ancestors") && uint32(cCtx.Int("ancestors")) < endMassif {
-						startMassif = endMassif - uint32(cCtx.Int("ancestors"))
-					}
-
-					err = replicator.ReplicateVerifiedUpdates(
-						context.Background(),
-						change.Tenant, startMassif, endMassif,
-					)
-					if err != nil {
-						errChan <- err
-					}
-				}(change, errChan)
+			concurency := min(len(changes), max(1, cCtx.Int("concurrency")))
+			for i := 0; i < len(changes); i += concurency {
+				err = replicateChanges(cCtx, cmd, changes[i:min(i+concurency, len(changes))], progress)
+				if err != nil {
+					return err
+				}
 			}
 
-			// the error channel is buffered enough for each tenant, so this will not get deadlocked
-			wg.Wait()
-			close(errChan)
-
-			var errs []error
-			for err := range errChan {
-				cmd.log.Infof(err.Error())
-				errs = append(errs, err)
-			}
-			if len(errs) > 0 {
-				return errs[0]
-			}
-			if len(changes) == 1 {
-				cmd.log.Infof("replication complete for tenant %s", changes[0].Tenant)
-			} else {
-				cmd.log.Infof("replication complete for %d tenants", len(changes))
-			}
 			return nil
 		},
 	}
 }
 
+// replicateChanges replicate the changes for the provided slice of tenants.
+// Paralelism is limited by breaking the total changes into smaller slices and calling this function
+func replicateChanges(cCtx *cli.Context, cmd *CmdCtx, changes []TenantMassif, progress Progresser) error {
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(changes)) // buffered so it doesn't block
+
+	for _, change := range changes {
+		wg.Add(1)
+		go func(change TenantMassif, errChan chan<- error) {
+			defer wg.Done()
+			defer progress.Completed()
+
+			retries := max(-1, cCtx.Int("retries"))
+			for {
+
+				replicator, err := NewVerifiedReplica(cCtx, cmd.Clone())
```

Comment:
minor: this is quite a large function, can we move things out of the for loop into a different function?

Reply:
I have refactored so that the initialisation of the replicator and start / end are broken out.

Breaking this up further obscures important details about how the go routines and the channels interact. I feel pretty strongly that this all belongs together